Copy extension link

cachandlerdev opened the pull request:
<!-- Thank you for submitting a Pull Request. Please:
* Read our Pull Request guidelines:
  https://github.com/microsoft/vscode/wiki/How-to-Contribute#pull-requests
* Associate an issue with the Pull Request.
* Ensure that the code is up-to-date with the `main` branch.
* Include a description of the proposed changes and how to test them.
-->
Hi, I added a context menu item for "Copy Extension Link" in accordance with #233915. I'm afraid I'm not quite sure how to test it on extensions from the vscode marketplace because they don't show up when running `./scripts/code.sh`, but the logic is copied from the "Copy" item above, and I believe it appropriately handles built-in extensions.

cachandlerdev committed:
fix: add "copy extension link" context item

cachandlerdev committed:
Merge branch 'microsoft:main' into copyExtensionLink

cachandlerdev committed:
fix: hide "copy extension link" on builtin extensions

cachandlerdev commented:
@microsoft-github-policy-service agree

SalerSimo commented:
@cachandlerdev If you go in the Extensions and you apply a filter some extension should show up.
I tried your code but i didnt get any url, what value is supposed to be wrote in the clipboard?

cachandlerdev commented:
Hi, I noticed that behavior as well, but I believe it's the case because those extensions are built-in by default and don't have links. If you look at the existing "Copy" code, you'll see that no link gets created for those extensions.

SalerSimo commented:
I managed to install an extension, you have to download the .vsix extension file from the marketplace and then go there:
![Screenshot 2024-11-19 224318](https://github.com/user-attachments/assets/c6491800-f18c-45f1-a74d-21b711be086f)
Even with this extension the function doesn't work, the parameter extension.url always seems to be 'undefined'

cachandlerdev committed:
Merge branch 'microsoft:main' into copyExtensionLink

cachandlerdev commented:
Ah, I see what you mean. However, this appears to be standard behavior for the "Copy" context menu item that I have not touched on this development version. I find this odd because when I run VSCode normally (not the development version), the URL is retrieved when clicking the "Copy" context menu item button.

SalerSimo commented:
Then instead of the "Copy" item you might copy the logic from "Copy Extension Id" and just add a prefix:
```
run: async (accessor: ServicesAccessor, extensionId: string) => {
      const clipboardService = accessor.get(IClipboardService);
      const extensionLinkPrefix = 'https://marketplace.visualstudio.com/items?itemName=';

      await clipboardService.writeText(extensionLinkPrefix + extensionId);
}
```
Something like this.

cachandlerdev committed:
fix: add "copy marketplace link" and fix when condition

cachandlerdev committed:
Merge branch 'microsoft:main' into copyExtensionLink

cachandlerdev commented:
Alright, I changed that code and corrected the `when:` condition to only display the menu item for extensions that aren't built-in (and as a result don't have marketplace links). The feature appears to be working now.

cachandlerdev committed:
Merge branch 'main' into copyExtensionLink

isidorn commented on src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts:
```diff
@@ -1572,6 +1572,21 @@ class ExtensionsContributions extends Disposable implements IWorkbenchContributi
+			},
+			run: async (accessor: ServicesAccessor, id: string) => {
+				const clipboardService = accessor.get(IClipboardService);
+				const extensionLinkPrefix = 'https://marketplace.visualstudio.com/items?itemName=';
```
You need to get the marketplace link from the productService and not to hard code it.

cachandlerdev replied:
I have now fixed that issue.

SalerSimo commented:
@cachandlerdev I think i figured out why using the logic from "Copy" doesn't work: you must be connected to internet, and the application built from the source code doesn't have internet access.
You can check this in the standard VS Code app by disabling your internet connection, reloading the window and then click on "Copy".

cachandlerdev commented:
I see what you mean. In that case, I'll revert to the prior logic (which should work because it's directly taken from the existing "Copy" item) but only display the context menu item when the user is connected to the internet.

cachandlerdev committed:
fix: revert to getting the marketplace link from the product serice

cachandlerdev committed:
fix: change name, simplify url retrieval, simplify marketplace check

cachandlerdev commented:
@sandy081 I see. I've made the changes as requested.
